src/pinnable_mapped_file.cpp

```diff
@@ -181,6 +226,79 @@ pinnable_mapped_file::pinnable_mapped_file(const std::filesystem::path& dir, boo
+   // non-sharable chainbase dbs using mapped mode are flushed to disk
+   // ----------------------------------------------------------------------------------
+   for (auto pmm : _instance_tracker)
+      pmm->save_database_file(true);
+
+   _file_mapped_region = bip::mapped_region(_file_mapping, bip::copy_on_write);
+   *((char*)_file_mapped_region.get_address()+header_dirty_bit_offset) = dirty; // set dirty bit in our memory mapping
+
+   _segment_manager = reinterpret_cast<segment_manager*>((char*)_file_mapped_region.get_address()+header_size);
+
+   // then clear the Soft-Dirty bits
+   // ------------------------------
+   pagemap_accessor pagemap;
+   if (pagemap.pagemap_supported()) {
+      if (!pagemap.clear_refs())
+         BOOST_THROW_EXCEPTION(std::system_error(make_error_code(db_error_code::clear_refs_failed)));
+      _instance_tracker.push_back(this); // so we can save dirty pages before another instance calls `clear_refs()`
+   }
+}
+
+// this is called after loading a snapshot, when database-map-mode was switched from `mapped_private` to
+// `mapped` to avoid running out of memory (because loading a snapshot causes all state pages to be modified).
+// This provides an opportunity to revert back to the `mapped_private` mode with it friendlier disk
+// usage characteristics.
+void pinnable_mapped_file::revert_to_private_mode() {
+   if (!_sharable)
+      return;
+
+   // do synchronous flush of all modified pages of our mapping
+   if(_file_mapped_region.flush(0, 0, false) == false)
+      std::cerr << "CHAINBASE: ERROR: syncing buffers failed" << '\n';
+   else {
+      // disk db file is up to date (with dirty bit set to true)
+      // we can kill the RW (`shared`) mapping and recreate a `copy_on_write` one.
+      _file_mapped_region = bip::mapped_region();
+      setup_copy_on_write_mapping();
+      _sharable = false;
+   }
+}
+
+// returns the number of pages flushed to disk
+size_t pinnable_mapped_file::check_memory_and_flush_if_needed() {
+   if (_non_file_mapped_mapping || _sharable || !_writable)
+      return 0;
+
+   // we are in `copy_on_write` mode.
+   static time_t check_time = 0;
+   constexpr int check_interval = 60; // seconds
+   constexpr size_t one_gb = 1ull << 30;
+
+   size_t written_pages {0};
+   const time_t current_time = time(NULL);
+   if(current_time >= check_time) {
+      check_time = current_time + check_interval;
+
+      size_t avail_ram_gb = get_available_ram() / one_gb;
+      if (avail_ram_gb <= 2) {
+         auto [src, sz] = get_region_to_save();
+         pagemap_accessor pagemap;
+         size_t offset = 0;
+         while(offset != sz && written_pages < (one_gb / sysconf(_SC_PAGESIZE))) {
+            size_t copy_size = std::min(_db_size_copy_increment,  sz - offset);
+            if (!pagemap.update_file_from_region({ src + offset, copy_size }, _file_mapping, offset, false, written_pages))
+               break;
+            offset += copy_size;
+         }
+      }
+   }
```

Comment:
Won't this write out the same 1GB of dirty pages over and over again? I would have thought you need a call to `clear_refs()` to reset the bits back to clean. But.. the `clear_refs` interface is all-or-nothing: so you will need to write out all dirty pages.

Reply:
yes, you are right, this doesn't help. I'll comment it out for now. I don't think the memory check I have is good enough to do a full dirty page write (I'm worried this might happen too often).

Reply:
Fixed (kinda)